File: cn/istarx/python/NO807_MaxIncreaseToKeepCitySkyline.py
class Solution:
    def maxIncreaseKeepingSkyline(self, grid: List[List[int]]) -> int:
        if not grid or len(grid) == 0 or len(grid) ==1:
            return 0
        N = len(grid)
        row = [-1] * N
        col = [-1] * N
        for i in range(N):
            for j in range(N):
                row[i] = max(row[i], grid[i][j])
                col[i] = max(col[i], grid[j][i])
        res = 0
        for i in range(N):
            for j in range(N):
                res = res + min(row[i], col[j]) - grid[i][j]
        return res

File: src/cn/istarx/java/NO807_MaxIncreaseToKeepCitySkyline.java
package cn.istarx.java;

public class NO807_MaxIncreaseToKeepCitySkyline {

    public static void main(String[] args) {
        int[][] grid = new int[][]{
                {3, 0, 8, 4},
                {2, 4, 5, 7},
                {9, 2, 6, 3},
                {0, 3, 1, 0}
        };
        System.out.println(maxIncreaseKeepingSkyline(grid));
    }

    private static int maxIncreaseKeepingSkyline(int[][] grid) {
        if (grid == null || grid.length == 0 || grid[0].length == 0 ||
                grid.length == 1 && grid[0].length == 1) {
            return 0;
        }
        int N = grid.length;
        int[] leftRight = new int[N];
        int[] topBottom = new int[N];
        for (int i = 0; i < N; i++) {
            for(int j = 0; j< N; j++) {
                leftRight[i] = Math.max(leftRight[i], grid[i][j]);
                topBottom[i] = Math.max(topBottom[i], grid[j][i]);
            }
        }

        int res = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                res += Math.min(leftRight[i], topBottom[j]) - grid[i][j];
            }
        }
        return res;
    }
}
